File: app/src/main/java/com/example/smartsave/model/Transaction.java
package com.example.smartsave.model;

import com.google.firebase.database.Exclude;
import com.google.firebase.database.IgnoreExtraProperties;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

@IgnoreExtraProperties
public class Transaction {
    private String id;

    private String description;
    private double amount;
    private String type;
    private double savingsCalculated;
    private long timestamp;
    private String currency;


    public Transaction() {
    }

    public Transaction(String description, double amount, String type, double savingsCalculated, long timestamp, String currency) {
        this.description = description;
        this.amount = amount;
        this.type = type;
        this.savingsCalculated = savingsCalculated;
        this.timestamp = timestamp;
        this.currency = currency;
    }

    public String getId() { return id; }
    public String getDescription() { return description; }
    public double getAmount() { return amount; }
    public String getType() { return type; }
    public double getSavingsCalculated() { return savingsCalculated; }
    public long getTimestamp() { return timestamp; }
    public String getCurrency() { return currency == null ? "EUR" : currency; } // Default currency if null

    public void setId(String id) { this.id = id; }
    public void setDescription(String description) { this.description = description; }
    public void setAmount(double amount) { this.amount = amount; }
    public void setType(String type) { this.type = type; }
    public void setSavingsCalculated(double savingsCalculated) { this.savingsCalculated = savingsCalculated; }
    public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
    public void setCurrency(String currency) { this.currency = currency; }

    @Exclude
    public String getFormattedDate(String format) {
        if (timestamp == 0L) return "Processing";
        SimpleDateFormat sdf = new SimpleDateFormat(format, Locale.getDefault());
        return sdf.format(new Date(timestamp));
    }

    @Exclude
    public String getFormattedDate() { // Overload with default format
        return getFormattedDate("dd MMM yyyy, HH:mm");
    }

    @Exclude
    public String getSavingsImpactForList() {
        switch (type != null ? type.toUpperCase() : "") {
            case "INCOME":
            case "EXPENSE":
                if (savingsCalculated > 0) {
                    return String.format(Locale.US, "+ %.2f %s", savingsCalculated, getCurrency());
                } else if (savingsCalculated < 0) {
                    return String.format(Locale.US, "%.2f %s", savingsCalculated, getCurrency());
                }
                return "";
            case "WITHDRAW":
                return String.format(Locale.US, "- %.2f %s", amount, getCurrency());
            case "SAVINGS_DEPOSIT":
            case "INTEREST_PAYMENT":
                return String.format(Locale.US, "+ %.2f %s", amount, getCurrency());
            default:
                return "";
        }
    }

    @Exclude
    public String getDisplayAmountForList() {
        String typeUpper = (type != null) ? type.toUpperCase() : "";

        switch (typeUpper) {
            case "INCOME":
            case "EXPENSE":
                return String.format(Locale.US, "Amount: %.2f %s", amount, getCurrency());
            case "WITHDRAW":
            case "SAVINGS_DEPOSIT":
            case "INTEREST_PAYMENT":
                return "";
            default:
                if (amount != 0) {
                    return String.format(Locale.US, "Amount: %.2f %s", amount, getCurrency());
                }
                return "";
        }
    }

    @Override
    public String toString() {
        return "Transaction{" +
                "id='" + id + '\'' +
                ", description='" + description + '\'' +
                ", amount=" + amount +
                ", type='" + type + '\'' +
                ", savingsCalculated=" + savingsCalculated +
                ", timestamp=" + timestamp +
                ", currency='" + currency + '\'' +
                '}';
    }
}

File: app/src/main/java/com/example/smartsave/data/models/Transaction.kt
package com.example.smartsave.model

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

data class Transaction(
    val id: String,             // UI-friendly ID (usually String)
    val timestamp: Long,        // Epoch milliseconds (UTC) for easy sorting/filtering
    val amount: Double,         // The primary amount for display
    val currency: String,       // The currency of the primary amount
    val description: String,    // UI-friendly description
    val type: String? = null    // Optional: "Credit", "Debit", or transaction_type from API
    // Add any other fields your UI specifically needs, derived or mapped
) {
    fun getFormattedDate(): String {
        if (timestamp == 0L) return "Invalid Date"
        // Example: "27 Oct 2023, 14:30"
        val sdf = SimpleDateFormat("dd MMM yyyy, HH:mm", Locale.getDefault())
        // Assuming timestamp is already UTC, format it for display in the user's local timezone
        sdf.timeZone = TimeZone.getDefault()
        return sdf.format(Date(timestamp))
    }

    fun getDisplayAmountForList(): String {
        // Example: "€123.45" or "$50.00"
        return String.format(Locale.getDefault(), "%.2f %s", amount, currency)
        // Consider using java.text.NumberFormat for locale-aware currency formatting
    }

    fun getSavingsImpactForList(): String {
        // This is specific to your SmartSave logic.
        // For example, if 'amount' is the saved portion:
        // if (amount > 0) return String.format(Locale.getDefault(), "+%.2f %s", amount, currency)
        // else if (amount < 0) return String.format(Locale.getDefault(), "%.2f %s", amount, currency) // Or handle negative savings
        // else return ""
        // For now, placeholder:
        return if (type == "Credit" && amount > 0) { // Simplified example: positive credit is savings
            String.format(Locale.getDefault(), "+%.2f %s", amount, currency)
        } else {
            "" // No savings impact or debit
        }
    }
    fun getSavingsCalculated(){

    }
}